core/src/main/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/MultiFieldStatsResults.java

```diff
@@ -0,0 +1,431 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Descriptive stats gathered per shard. Coordinating node computes final pearson product coefficient
+ * based on these descriptive stats
+ *
+ * @internal
+ */
+class MultiFieldStatsResults implements Streamable {
+    /** count of observations (same number of observations per field) */
+    protected long docCount;
+    /** per field sum of observations */
+    protected HashMap<String, Double> fieldSum;
+    /** counts */
+    protected HashMap<String, Long> counts;
+    /** mean values */
+    protected HashMap<String, Double> means;
+    /** variance values */
+    protected HashMap<String, Double> variances;
+    /** skewness values */
+    protected HashMap<String, Double> skewness;
+    /** kurtosis values */
+    protected HashMap<String, Double> kurtosis;
+    /** covariance values */
+    protected HashMap<String, HashMap<String, Double>> covariances;
+    /** pearson product correlation coefficients */
+    protected HashMap<String, HashMap<String, Double>> correlation;
+
+    /** cache flag to avoid recompute if nothing has been added */
+    private boolean dirty;
+
+    /**
+     * Base ctor
+     */
+    public MultiFieldStatsResults() {
+        docCount = 0;
+        dirty = true;
+        counts = new HashMap<>();
+        fieldSum = new HashMap<>();
+        means = new HashMap<>();
+        skewness = new HashMap<>();
+        kurtosis = new HashMap<>();
+        covariances = new HashMap<>();
+        variances = new HashMap<>();
+    }
+
+    /**
+     * Returns covariance between two fields
+     */
+    public double getCovariance(String fieldX, String fieldY) {
+        if (fieldX.equals(fieldY)) {
+            return variances.get(fieldX);
+        } else if (variances.containsKey(fieldX) == false && variances.containsKey(fieldY) == false) {
+            return Double.NaN;
+        }
+        return getValFromUpperTriangularMatrix(covariances, fieldX, fieldY);
+    }
+
+    /**
+     * Returns the correlation coefficient between two fields
+     */
+    public double getCorrelation(String fieldX, String fieldY) {
+        if (fieldX.equals(fieldY)) {
+            return 1.0;
+        } else if (correlation.containsKey(fieldX) == false && correlation.containsKey(fieldY) == false) {
+            return Double.NaN;
+        }
+        return getValFromUpperTriangularMatrix(correlation, fieldX, fieldY);
+    }
+
+    /**
+     * Gets the value for two fields in an upper triangular matrix, regardless of row col location.
+     */
+    private double getValFromUpperTriangularMatrix(HashMap<String, HashMap<String, Double>> map, String fieldX, String fieldY) {
+        if (map.containsKey(fieldX)) {
+            if (map.get(fieldX).containsKey(fieldY)) {
+                return map.get(fieldX).get(fieldY);
+            } else {
+                return map.get(fieldY).get(fieldX);
```

Comment:
e.g. from above comment, if `fieldX` exists, `fieldX` does not contain `fieldY`, and `fieldY` does not exist... it'll blowup.

But perhaps that's an impossible situation for this code to get into?